### DISCUSSION TITLE
Can you change the input device?

### DISCUSSION THREAD

**Discussion by alankent** [katspaugh/wavesurfer.js#3164] [Q&A]
I was looking through MediaDevices and methods like enumerateDevices() to find all the video and audio input devices, created a drop down menu, then realized I did not know how to tell wavesurfer how to change input device. In particular, I think https://github.com/katspaugh/wavesurfer.js/blob/d82ed8b43177da30eea44e95830bb019638bd9e3/src/plugins/record.ts#L78 means I cannot specify it. (You would use `audio: {deviceId: xxx}` instead of `audio: true`.) Is this correct? Any suggested work arounds?

**Comment by katspaugh** [comment]
Feel free to make a pull request to allow this parametrization.

**Comment by alankent** [comment]
Thanks. I will investigate a bit more. I have multiple microphones, but Chrome is only returning me one. (I am getting multiple camera devices.) It may be related to security and device access, so I will try and get audio integrated then come back and look at passing it through. I think its just a matter of adding deviceId or similar to the create options list (if present, replace 'true' with the device id). I can handle that level of contribution! ;-)

**Comment by alankent** [comment]
Hmm. I need to work out the flow a bit better. I want to show the record level without starting the recording yet. This will also trigger the permissions prompt, after which I am now seeing a full list of audio devices. So I really want an "enable audio" button and a "record" button. The first triggers the prompt and starts showing the graph, the second starts the actual recording...   I have to disappear for a while, but making progress!

**Reply by katspaugh** [reply]
Sounds good! Let me know if you have any questions along the way!

**Comment by alankent** [comment]
I am also having trouble turning the microphone off.  Chrome is showing it as "still recording" even though I stop() all the tracks. I am guessing because I am controlling the media on/off (trying to implement "enable/disable" separate to "record"), but wavesurfer also asks for the media device which turns it back on afterwards (maybe). The camera is working fine, its just audio causing a problem. Fun! 

<img width="29" alt="image" src="https://github.com/katspaugh/wavesurfer.js/assets/5702163/24b32693-a9da-4c08-91e2-d7528160cdd2">

I almost want to control the media stream in my code (I am turning the camera on/off as well). I want to turn the mic on when and see the volume meter, then later hit the "record" button to start/stop recording. I am specifying the deviceId for the camera and microphone. Still thinking it through. 

<img width="276" alt="image" src="https://github.com/katspaugh/wavesurfer.js/assets/5702163/1eda24d7-e0f9-4110-bd9c-592de5e6ecc4">

**Comment by alankent** [comment]
Oh, and the above list of devices only gets fleshed out after the user grants permission to access devices. So I ask for access to audio and video when they click the settings button, so the menu will get fleshed out. So its all fun and games synchronizing the calls.

**Comment by alankent** [comment]
I was having so much trouble getting the interaction exactly as I wanted (including with the video camera feeding into a separate library), I ended up just using ChatGPT to generate code to call the recording API directly. It ended up less code. Next stage is viewing the audio clips (which is what I was originally interested in wave surfer for!)
